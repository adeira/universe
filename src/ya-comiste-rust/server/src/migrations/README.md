These migrations are written in Rust on purpose. It allows you to:

- do whatever you could do from Rust code (not limited to some YAML or AQL syntax)
- migrate not only structure but manipulate data as well
- write integration tests for these migrations

As a disadvantage they might be more verbose.
