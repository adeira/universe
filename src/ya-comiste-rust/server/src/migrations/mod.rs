#![allow(non_snake_case)]

use crate::arangodb::ConnectionPool;
use arangors::document::options::InsertOptions;
use arangors::ClientError;
use futures::future::BoxFuture;

mod utils;

mod v0__migrations_init;
mod v1__create_collection_users_sessions;
mod v2__create_document_user_anonymous;
mod v3__create_document_user_admin;
mod v4__create_collection_products;
mod v5__create_collection_images;

#[derive(serde::Serialize, serde::Deserialize)]
struct MigrationRecord {
    _key: String,
}

// This migrations should be written to be idempotent: not to fail when they are run repeatedly.
pub async fn migrate(pool: &ConnectionPool) {
    let db = pool.db().await;

    // https://users.rust-lang.org/t/error-distinct-uses-of-impl-trait-result-in-different-opaque-types/46862/2
    // https://users.rust-lang.org/t/how-to-handle-a-vector-of-async-function-pointers/39804
    // https://users.rust-lang.org/t/how-to-store-async-functions-in-a-vector/51630
    type MigrationFunction = fn(
        &arangors::Database<arangors::client::reqwest::ReqwestClient>,
    ) -> BoxFuture<'_, Result<(), ClientError>>;

    // TODO: hide behind some macro (?)
    let migrations: Vec<(&str, MigrationFunction)> = vec![
        ("v0__migrations_init", |db| {
            Box::pin(v0__migrations_init::migrate(&db))
        }),
        ("v1__create_collection_users_sessions", |db| {
            Box::pin(v1__create_collection_users_sessions::migrate(&db))
        }),
        ("v2__create_document_user_anonymous", |db| {
            Box::pin(v2__create_document_user_anonymous::migrate(&db))
        }),
        ("v3__create_document_user_admin", |db| {
            Box::pin(v3__create_document_user_admin::migrate(&db))
        }),
        ("v4__create_collection_products", |db| {
            Box::pin(v4__create_collection_products::migrate(&db))
        }),
        ("v5__create_collection_images", |db| {
            Box::pin(v5__create_collection_images::migrate(&db))
        }),
    ];

    log::info!("Checking pending database migrations");
    for (migration_name, migration) in &migrations {
        if let Ok(migrations) = db.collection("migrations").await {
            if migrations
                .document::<MigrationRecord>(migration_name)
                .await
                .is_ok()
            {
                // the migration was already applied before - skip
                continue;
            }
        }

        let migration_result = migration(&db).await;
        match migration_result {
            Ok(_) => {
                // migration was successful - let's remember it so we don't run it again later
                if let Ok(migrations) = db.collection("migrations").await {
                    let result_future = migrations.create_document(
                        MigrationRecord {
                            _key: migration_name.to_string(),
                        },
                        InsertOptions::builder().return_new(true).build(),
                    );
                    match result_future.await {
                        Ok(_) => log::info!("{} ✅", migration_name),
                        Err(e) => {
                            log::error!("{} ❌", migration_name);
                            panic!("{:?}", e)
                        }
                    }
                }
            }
            Err(e) => {
                log::error!("{} ❌", migration_name);
                panic!("{:?}", e)
            }
        }
    }
}
