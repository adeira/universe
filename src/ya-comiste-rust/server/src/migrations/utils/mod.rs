use arangors::collection::options::{CreateOptions, CreateParameters};
use arangors::collection::CollectionType;
use arangors::document::options::InsertOptions;
use arangors::index::Index;
use arangors::ClientError;
use serde::{Deserialize, Serialize};

pub(in crate::migrations) trait ArangoDocument {
    /// This key makes sure we don't create the same document twice by accident
    /// when running the migrations repeatedly.
    fn idempotency_key(&self) -> &str;
}

pub(in crate::migrations) async fn create_collection(
    db: &arangors::Database<arangors::client::reqwest::ReqwestClient>,
    collection_name: &str,
    collection_type: &CollectionType,
    json_schema: &Option<serde_json::Value>,
) -> Result<(), ClientError> {
    match db.collection(collection_name).await {
        Ok(_) => {
            // collection already exists, nothing to do
            Ok(())
        }
        Err(_) => {
            // collection doesn't exist yet, let's create it
            let create_options = match json_schema {
                Some(json_schema) => CreateOptions::builder()
                    .name(collection_name)
                    .collection_type(*collection_type)
                    .schema(json_schema.to_owned())
                    .build(),
                None => CreateOptions::builder()
                    .name(collection_name)
                    .collection_type(*collection_type)
                    .build(),
            };
            let result = db
                .create_collection_with_options(create_options, CreateParameters::default())
                .await;

            match result {
                Ok(_) => Ok(()),
                Err(e) => Err(e),
            }
        }
    }
}

pub(in crate::migrations) async fn create_document<T>(
    db: &arangors::Database<arangors::client::reqwest::ReqwestClient>,
    collection_name: &str,
    document: T,
) -> Result<(), ClientError>
where
    T: Serialize + for<'de> Deserialize<'de> + ArangoDocument,
{
    let collection = db.collection(&collection_name).await.unwrap(); // TODO
    match collection.document::<T>(document.idempotency_key()).await {
        Ok(_) => {
            // user already exists
            Ok(())
        }
        Err(_) => {
            // user doesn't exist yet, let's create it
            let result = collection
                .create_document(document, InsertOptions::builder().return_new(true).build())
                .await;

            match result {
                Ok(_) => Ok(()),
                Err(e) => Err(e),
            }
        }
    }
}

pub(in crate::migrations) async fn create_index(
    db: &arangors::Database<arangors::client::reqwest::ReqwestClient>,
    collection_name: &str,
    index: &Index,
) -> Result<(), ClientError> {
    match db.create_index(collection_name, &index).await {
        Ok(_) => Ok(()),
        Err(e) => Err(e),
    }
}
