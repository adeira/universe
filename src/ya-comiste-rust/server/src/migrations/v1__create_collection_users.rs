use crate::migrations::utils::create_collection;
use arangors::collection::CollectionType;
use arangors::ClientError;

pub async fn migrate(
    db: &arangors::Database<arangors::client::reqwest::ReqwestClient>,
) -> Result<(), ClientError> {
    let collection_name = "users";
    let json_schema = serde_json::from_str(
        r##"
        {
          "message": "",
          "level": "strict",
          "rule": {
            "type": "object",
            "additionalProperties": false,
            "required": [
              "type"
            ],
            "properties": {
              "type": {
                "type": "string",
                "enum": [
                  "anonymous",
                  "regular",
                  "admin"
                ]
              },
              "google": {
                "type": "object",
                "additionalProperties": true,
                "required": [
                  "sub"
                ],
                "properties": {
                  "sub": {
                    "type": "string"
                  }
                }
              }
            }
          }
        }
        "##,
    )
    .unwrap();

    create_collection(
        &db,
        &collection_name,
        &CollectionType::Document,
        &json_schema,
    )
    .await
}
