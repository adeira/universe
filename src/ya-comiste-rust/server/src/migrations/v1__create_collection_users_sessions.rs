use crate::migrations::utils::{create_collection, create_index};
use arangors::collection::CollectionType;
use arangors::index::{Index, IndexSettings};
use arangors::ClientError;

pub async fn migrate(
    db: &arangors::Database<arangors::client::reqwest::ReqwestClient>,
) -> Result<(), ClientError> {
    // 1. create `users` table
    create_collection(
        &db,
        "users",
        &CollectionType::Document,
        &serde_json::from_str(
            r##"
            {
              "message": "",
              "level": "strict",
              "rule": {
                "type": "object",
                "additionalProperties": false,
                "required": [
                  "type"
                ],
                "properties": {
                  "type": {
                    "type": "string",
                    "enum": [
                      "anonymous",
                      "regular",
                      "admin"
                    ]
                  },
                  "google": {
                    "type": "object",
                    "additionalProperties": true,
                    "required": [
                      "sub"
                    ],
                    "properties": {
                      "sub": {
                        "type": "string"
                      }
                    }
                  }
                }
              }
            }
            "##,
        )
        .unwrap(),
    )
    .await?;

    // 2. create `sessions` table
    create_collection(
        &db,
        "sessions",
        &CollectionType::Document,
        &serde_json::from_str(
            r##"
            {
              "message": "",
              "level": "strict",
              "rule": {
                "type": "object",
                "additionalProperties": false,
                "required": ["last_access", "type"],
                "properties": {
                  "last_access": {
                    "type": "string"
                  },
                  "type": {
                    "type": "string",
                    "enum": ["mobile", "webapp"]
                  }
                }
              }
            }
            "##,
        )
        .unwrap(),
    )
    .await?;

    // 3. create edge between `users` and `sessions`
    create_collection(&db, "user_sessions", &CollectionType::Edge, &None).await?;

    // 4. create TTL index for the sessions
    create_index(
        &db,
        "sessions",
        &Index::builder()
            .name("last_access_mobile_ttl_index") // required in migrations for idempotency
            .fields(vec!["last_access".to_string()])
            .settings(IndexSettings::Ttl {
                // Mobile session token is long-lived (one year = 365 * 24 * 60 * 60)
                expire_after: 31_536_000,
            })
            .build(),
    )
    .await
}
