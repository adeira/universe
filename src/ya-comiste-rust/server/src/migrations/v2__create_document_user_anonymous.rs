use crate::migrations::utils::create_document;
use arangors::ClientError;

#[derive(serde::Serialize, serde::Deserialize)]
struct User {
    _key: String,
    r#type: String,
}

pub async fn migrate(
    db: &arangors::Database<arangors::client::reqwest::ReqwestClient>,
) -> Result<(), ClientError> {
    let collection_name = "users";
    create_document(
        &db,
        &collection_name,
        User {
            _key: String::from("1"),
            r#type: String::from("anonymous"),
        },
        "1",
    )
    .await
}
