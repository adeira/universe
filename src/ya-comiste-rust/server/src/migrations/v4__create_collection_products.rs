use crate::migrations::utils::{create_collection, create_document, ArangoDocument};
use arangors::collection::CollectionType;
use arangors::ClientError;

#[derive(serde::Serialize, serde::Deserialize)]
struct ProductUnit {
    _key: String,
    en_US: String,
    es_MX: String,
}

impl ArangoDocument for ProductUnit {
    fn idempotency_key(&self) -> &str {
        &self._key
    }
}

pub async fn migrate(
    db: &arangors::Database<arangors::client::reqwest::ReqwestClient>,
) -> Result<(), ClientError> {
    let json_schema_products = serde_json::from_str(
        r##"
        {
          "message": "",
          "level": "strict",
          "rule": {
            "type": "object",
            "additionalProperties": false,
            "required": ["images", "unit_label", "active", "created", "updated", "price", "translations"],
            "properties": {
              "images": {
                "type": "array",
                "items": {
                  "type": "string"
                }
              },
              "unit_label": {
                "type": "string"
              },
              "active": {
                "type": "boolean"
              },
              "created": {
                "type": "string"
              },
              "updated": {
                "type": "string"
              },
              "price": {
                "type": "object",
                "additionalProperties": false,
                "required": ["currency", "unit_amount"],
                "properties": {
                  "currency": {
                    "type": "string",
                    "enum": ["mxn"]
                  },
                  "unit_amount": {
                    "type": "number"
                  }
                }
              },
              "translations": {
                "type": "object",
                "additionalProperties": false,
                "properties": {
                  "en_US": {
                    "type": "object",
                    "additionalProperties": false,
                    "required": ["name", "description"],
                    "properties": {
                      "name": {
                        "type": "string"
                      },
                      "description": {
                        "type": "string"
                      }
                    }
                  },
                  "es_MX": {
                    "type": "object",
                    "additionalProperties": false,
                    "required": ["name", "description"],
                    "properties": {
                      "name": {
                        "type": "string"
                      },
                      "description": {
                        "type": "string"
                      }
                    }
                  }
                }
              }
            }
          }
        }
        "##,
    )
    .unwrap();

    create_collection(
        &db,
        "products",
        &CollectionType::Document,
        &json_schema_products,
    )
    .await?;

    create_collection(&db, "product_units", &CollectionType::Document, &None).await?;

    create_document(
        &db,
        "product_units",
        ProductUnit {
            _key: String::from("piece"),
            en_US: String::from("piece"),
            es_MX: String::from("pieza"),
        },
    )
    .await
}
