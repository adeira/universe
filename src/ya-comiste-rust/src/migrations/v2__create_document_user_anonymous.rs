use crate::migrations::utils::{create_document, ArangoDocument};
use arangors::ClientError;

#[derive(serde::Serialize, serde::Deserialize)]
struct User {
    _key: String,
    r#type: String,
}

impl ArangoDocument for User {
    fn idempotency_key(&self) -> &str {
        &self._key
    }
}

pub async fn migrate(
    db: &arangors::Database<arangors::client::reqwest::ReqwestClient>,
) -> Result<(), ClientError> {
    let collection_name = "users";
    create_document(
        &db,
        &collection_name,
        User {
            _key: String::from("1"),
            r#type: String::from("anonymous"),
        },
    )
    .await
}
