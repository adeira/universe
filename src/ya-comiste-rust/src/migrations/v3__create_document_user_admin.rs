use crate::migrations::utils::{create_document, ArangoDocument};
use arangors::ClientError;

#[derive(serde::Serialize, serde::Deserialize)]
struct User {
    _key: String,
    r#type: String,
    google: UserGoogleClaims,
}

#[derive(serde::Serialize, serde::Deserialize)]
struct UserGoogleClaims {
    iss: String,
    sub: String,
    azp: String,
    aud: String,
    iat: usize,
    exp: usize,
    email: String,
    email_verified: bool,
    name: String,
    picture: String,
    given_name: String,
    family_name: String,
    locale: String,
}

impl ArangoDocument for User {
    fn idempotency_key(&self) -> &str {
        &self._key
    }
}

pub async fn migrate(
    db: &arangors::Database<arangors::client::reqwest::ReqwestClient>,
) -> Result<(), ClientError> {
    let user_google_claims = UserGoogleClaims {
        iss: String::from("https://accounts.google.com"),
        sub: String::from("108269453578187886435"),
        azp: String::from("245356693889-63qeuc6183hab6be342blikbknsvqrhk.apps.googleusercontent.com"),
        aud: String::from("245356693889-63qeuc6183hab6be342blikbknsvqrhk.apps.googleusercontent.com"),
        iat: 1610381356,
        exp: 1610384956,
        email: String::from("mrtnzlml@gmail.com"),
        email_verified: true,
        name: String::from("Martin Zlámal"),
        picture: String::from("https://lh3.googleusercontent.com/a-/AOh14Gg_ZmWYHHMei_AQdckVaPqLikBkNYJONyyR0QqWfWs=s96-c"),
        given_name: String::from("Martin"),
        family_name: String::from("Zlámal"),
        locale: String::from("en")
    };

    let user = User {
        _key: String::from("2"),
        r#type: String::from("admin"),
        google: user_google_claims,
    };

    create_document(&db, "users", user).await
}
