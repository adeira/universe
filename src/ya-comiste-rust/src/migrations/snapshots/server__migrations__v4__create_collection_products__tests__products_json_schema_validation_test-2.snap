---
source: server/src/migrations/v4__create_collection_products.rs
expression: "schema.validate(&json!({ }))"
---
{
  "errors": [
    {
      "code": "required",
      "title": "This property is required",
      "path": "/images"
    },
    {
      "code": "required",
      "title": "This property is required",
      "path": "/unit_label"
    },
    {
      "code": "required",
      "title": "This property is required",
      "path": "/is_published"
    },
    {
      "code": "required",
      "title": "This property is required",
      "path": "/visibility"
    },
    {
      "code": "required",
      "title": "This property is required",
      "path": "/created"
    },
    {
      "code": "required",
      "title": "This property is required",
      "path": "/updated"
    },
    {
      "code": "required",
      "title": "This property is required",
      "path": "/price"
    },
    {
      "code": "required",
      "title": "This property is required",
      "path": "/translations"
    }
  ],
  "missing": []
}
